### DISCUSSION TITLE
bold, italics, underline

### DISCUSSION THREAD

**Discussion by lbr991** [pymupdf/PyMuPDF#2641] [Looking for help] [answered]
Is there a way to apply bold/italics/underline on a font or do we have to switch fonts?

**Comment by JorjMcKie** [comment]
PyMuPDF offers multiple ways to output text:
* `page.insert_text()`, `page.insert_textbox()`: you have to switch fonts and compute the right continuing text position youself.
* `fitz.TextWriter` together with its methods: You also have to switch fonts, but there are elegant ways to find the write point to continue text with a different font.

In any above case, "underline" is no font option. You always must use a draw method or an underline annotation.

A third option is using `fitz.Story`. You must prepare a HTML source for your text. You can use all features offered by HTML5 syntax - including CSS styles.
Outputting a stroy is automatic, including line breaks and page breaks, font switching, colors, images etc. Please look up the respective documentation.

**Reply by lbr991** [reply]
Thanks!
Is there a way to identify bold, italic and underlined words in a pdf? 
I tried with the following but it won't print anything (the pdf contains bold words):

```
# Open the PDF file
pdf = fitz.open('input.pdf')

# Loop through each page
for page_num in range(len(pdf)):
    page = pdf[page_num]

    # Get text as a dictionary, including its font attributes
    text_info = page.get_text("dict")

    # Loop through the blocks of text
    for block in text_info['blocks']:
        # Check if the block contains text
        if block['type'] == 0:
            # Loop through the lines in the block
            for line in block['lines']:
                # Loop through the spans in the line
                for span in line['spans']:
                    # Check if the font is bold
                    if span['flags'] == 2:  # The value 2 generally signifies bold text
                        print(span['text'])

```

**Reply by JorjMcKie** [reply]
Again: underlining cannot be determined by looking at the font.

Otherwise, `span["flags"]` is a bit field, so you must test via `span["flags"] & 2**n` to test for one of the font weights or font styles - **_not_** for equality as you did.
In addition, there exist ways to "simulate" bold with a regular font by outputting the same text twice with a small displacement (and similar techniques). In thess cases you can test bits as long as you want and you still won't detect this.
This done often when the PDF creator wanted to avoid embedding another font file (to save file size).

**Reply by lbr991** [reply]
Oh interesting. Is there a way to detect the latter case?

**Reply by JorjMcKie** [reply]
Check if the same text appears twice on extraction, the second mostly overlapping the first bbox.
But other methods can even evade this by giving the text characters a fat border line - not detectable.

**Reply by lbr991** [reply]
Thanks!
Can you explain what n is in `span["flags"] & 2**n`? I got it to print out the bold text with `span["flags"] & 2**4` but could it be a different n in another case?

**Reply by JorjMcKie** [reply, marked as answer]
Look at the documentation [here](https://pymupdf.readthedocs.io/en/latest/textpage.html#structure-of-dictionary-outputs):
![grafik](https://github.com/pymupdf/PyMuPDF/assets/8290722/619b4b0d-8779-4593-b2c0-c1f001924819)

**Reply by lbr991** [reply]
> * `page.insert_text()`, `page.insert_textbox()`: you have to switch fonts and compute the right continuing text position youself.
> * `fitz.TextWriter` together with its methods: You also have to switch fonts, but there are elegant ways to find the write point to continue text with a different font.

The issue is that I want the text to wrap. insert_textbox() has the wrapping capability but fitz.TextWriter makes it easier to switch the fonts. 
How can I switch the fonts with insert_textbox()? Or should I use a combination of both?

**Reply by JorjMcKie** [reply]
The problem with `page.insert_textbox()` is that you cannot easily find the exact location where the insertion ends.
So if you need switching fonts, `TextWriter` is your way to go. Its `.fill_textbox()` method is comparable plus it returns the point where the text ends. This can be used as the starting point for the text in a different font.
Mixing the two approaches is not possible.

**Reply by lbr991** [reply]
Sounds tricky, does fill_textbox() support wrapping of text? Looked at the docs but did not see this.

**Reply by JorjMcKie** [reply]
Yes it does. Also look at the examples [here](https://github.com/pymupdf/PyMuPDF-Utilities/tree/master/textwriter).

**Reply by lbr991** [reply]
I'm using the following code and I get overlapping lines (no wrap), what am I doing wrong?
```
import fitz

pdf_name = 'input.pdf'
normal_font = fitz.Font("Helvetica")
bold_font = fitz.Font("Helvetica-Bold")

# Load original PDF
original_pdf = fitz.open(pdf_name)
output_pdf_path = 'out.pdf'
pdf_writer = fitz.open()

# Helper function to check if a span is bold
def is_bold(span):
    return span['flags'] & 2**4  # Bitwise AND to check if the 2nd bit is on, generally signifies bold text

for i in range(len(original_pdf)):
    orig_page = original_pdf.load_page(i)
    new_page = pdf_writer.new_page(width=orig_page.rect.width, height=orig_page.rect.height)
    text_dict = orig_page.get_text("dict")
    
    twriter = fitz.TextWriter(new_page.rect)

    for block in text_dict['blocks']:
        if block['type'] == 0:  # This is a text block
            bbox = fitz.Rect(block['bbox'])

            for line in block['lines']:
                for span in line['spans']:
                    original_text = span['text']
                    font_size = span['size']

                    # Use fill_textbox to adjust and fit the text inside the bounding box
                    if is_bold(span):
                        twriter.fill_textbox(bbox, original_text, fontsize=font_size, font=bold_font, warn=True)
                    else:
                        twriter.fill_textbox(bbox, original_text, fontsize=font_size, font=normal_font, warn=True)

    # Write the text to the new page
    twriter.write_text(new_page)

pdf_writer.save(output_pdf_path)
```

**Reply by JorjMcKie** [reply]
After writing the span text get the `last_point` property and use it as the starting point (`pos` parameter) for the next `fill_textbox()`.

**Reply by lbr991** [reply]
Can't figure it out, the example (https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/textwriter/textwriter-textbox.py) has line breaks. I'm passing a paragraph each time (no line breaks), insert_textbox() works fine but I don't know how to make text wrap with fill_textbox(). Maybe I misunderstand something?

**Reply by lbr991** [reply]
```
import fitz

pdf_name = 'input.pdf'
normal_font = fitz.Font("Helvetica")
bold_font = fitz.Font("Helvetica-Bold")

# Load original PDF
original_pdf = fitz.open(pdf_name)
output_pdf_path = 'out.pdf'
pdf_writer = fitz.open()

# Helper function to check if a span is bold
def is_bold(span):
    return span['flags'] & 2**4

# Helper function to calculate required height for a textbox
def calculate_required_height(text, fontsize, width):
    lines = text.split("\n")
    required_height = len(lines) * fontsize  # A simple approximation, can be adjusted
    return required_height

for i in range(len(original_pdf)):
    orig_page = original_pdf.load_page(i)
    new_page = pdf_writer.new_page(width=orig_page.rect.width, height=orig_page.rect.height)
    twriter = fitz.TextWriter(new_page.rect)

    y_offset = 0
    blocks = orig_page.get_text("dict")['blocks']

    for block in blocks:
        if block['type'] == 0:  # This is a text block
            paragraph = ""
            bold_markers = []
            fontsize = block['lines'][0]['spans'][0]['size']

            start_index = 0
            for line in block['lines']:
                for span in line['spans']:
                    paragraph += span['text'] + " "
                    if is_bold(span):
                        bold_markers.append((start_index, start_index + len(span['text'])))
                    start_index += len(span['text']) + 1

            fill_rect = fitz.Rect(block['bbox'])
            required_height = calculate_required_height(paragraph, fontsize, fill_rect.width)
            
            # Adjust the bounding box
            if required_height > fill_rect.height:
                diff = required_height - fill_rect.height
                fill_rect.y1 += diff
                y_offset += diff

            # Construct the mixed-font paragraph
            start, end = 0, 0
            for marker in bold_markers:
                end = marker[0]
                twriter.fill_textbox(fill_rect, paragraph[start:end], font=normal_font, fontsize=fontsize, warn=True)
                start = end
                end = marker[1]
                twriter.fill_textbox(fill_rect, paragraph[start:end], font=bold_font, fontsize=fontsize, warn=True)
                start = end
            
            # Add the remaining part
            twriter.fill_textbox(fill_rect, paragraph[start:], font=normal_font, fontsize=fontsize, warn=True)

    twriter.write_text(new_page)

pdf_writer.save(output_pdf_path)

```
Made some progress but now there are overlapping lines when the paragraph contains mix of bold and normal font. Also a weird indentation at the 2nd line (had it on the 1st line using some other code) and the spacing between the words of the paragraph is inconsistent.  

![image](https://github.com/pymupdf/PyMuPDF/assets/136107003/eddc8274-f4e6-4760-b054-5b16ea000b6a)

**Reply by lbr991** [reply]
@JorjMcKie Any ideas on how to fix this?

**Reply by JorjMcKie** [reply]
You still seem to not use the `last_point` attribute of TextWriter. This is the bottom-right of the last character written. Method `fill_textbox()` has a keyword parameter `pos`. I would use it by providing it the value of `lat_point`.
What does "inconsistent inter-word spacing" mean?
